src/pep257.py

```diff
@@ -679,73 +678,496 @@ def __getattr__(self, item):
+RunConfiguration = namedtuple('RunConfiguration',
+                              ('explain', 'source', 'debug',
+                               'verbose', 'count'))
 
-    Example
-    -------
-    >>> sorted(collect(['non-dir.txt', './'],
-    ...                match=lambda name: name.endswith('.py')))
-    ['non-dir.txt', './pep257.py', './setup.py', './test_pep257.py']
+
+class IllegalConfiguration(Exception):
+    """An exception for illegal configurations."""
+
+    pass
+
+
+# Check configuration - used by the ConfigurationParser class.
+CheckConfiguration = namedtuple('CheckConfiguration',
+                                ('checked_codes', 'match', 'match_dir'))
+
+
+def check_initialized(method):
+    """Check that the configuration object was initialized."""
+    def _decorator(self, *args, **kwargs):
+        if self._arguments is None or self._options is None:
+            raise RuntimeError('using an uninitialized configuration')
+        return method(self, *args, **kwargs)
+    return _decorator
+
+
+class ConfigurationParser(object):
+    """Responsible for parsing configuration from files and CLI.
+
+    There are 2 types of configurations: Run configurations and Check
+    configurations.
+
+    Run Configurations:
+    ------------------
+    Responsible for deciding things that are related to the user interface,
+    e.g. verbosity, debug options, etc.
+    All run configurations default to `False` and are decided only by CLI.
+
+    Check Configurations:
+    --------------------
+    Configurations that are related to which files and errors will be checked.
+    These are configurable in 2 ways: using the CLI, and using configuration
+    files.
+
+    Configuration files are nested within the file system, meaning that the
+    closer a configuration file is to a checked file, the more relevant it will
+    be. For instance, imagine this folder structure:
+
+    A
+    +-- tox.ini: sets `select=D100`
+    +-- B
+        +-- foo.py
+        +-- tox.ini: sets `add-ignore=D100`
+
+    Then `foo.py` will not be checked for `D100`.
+    The configuration build algorithm is described in `self._get_config`.
+
+    Note: If any of `MUTUALLY_EXCLUSIVE_OPTIONS` was selected in the CLI, all
+    configuration files will be ignored and each file will be checked for
+    the error codes supplied in the CLI.
 
     """
-    for name in names:  # map(expanduser, names):
-        if os.path.isdir(name):
-            for root, dirs, filenames in os.walk(name):
-                # Skip any dirs that do not match match_dir
-                dirs[:] = [dir for dir in dirs if match_dir(dir)]
-                for filename in filenames:
-                    if match(filename):
-                        yield os.path.join(root, filename)
+
+    CONFIG_FILE_OPTIONS = ('convention', 'select', 'ignore', 'add-select',
+                           'add-ignore', 'match', 'match-dir')
+    MUTUALLY_EXCLUSIVE_OPTIONS = ('ignore', 'select', 'convention')
```

Comment:
These options ARE mutually exclusive, but I'm not sure if that's their defining characteristic. Perhaps it should be `BASE_ERROR_SELECTION_OPTIONS` or something similar?